File: dataStructure_algorithm/chap02_string/2-4.py
from collections import defaultdict
from typing import List


# 그룹 애너그램(anagram)

input_value = ['eat', 'repaid', 'paired', 'tea', 'bat']

# 방법 1) 정렬 후 해시테이블을 이용하여 구하기


def groupAnagram1(strs: List[str]) -> List[List[str]]:
    hashmap = defaultdict(list)

    for s in strs:
        hashmap[''.join(sorted(s))].append(s)

    return hashmap.values()


print(groupAnagram1(input_value))

# 방법 2) 문자 카운트와 해시 테이블 이용


def groupAnagram2(strs: List[str]) -> List[List[str]]:
    hashmap = defaultdict(list)

    for s in strs:
        # 모든 소문자들의 개수를 담을 배열 생성
        count = [0] * 26

        for ch in s:
            count[ord(ch) - ord('a')] += 1

        # 해시 테이블의 키는 리스트는 불가능하지만 튜플은 가능
        hashmap[tuple(count)].append(s)

    return hashmap.values()


print(groupAnagram2(input_value))

File: dataStructure_algorithm/chap02_string/2-4.js
const input_value = ['eat', 'repaid', 'paired', 'tea', 'bat'];

// 방법 1)

function groupAnagram1(strs) {
  hashmap = new Map();

  for (const item of strs) {
    let sorted = [...item].sort().join('');

    if (hashmap.has(sorted)) {
      hashmap.get(sorted).push(item);
    } else {
      hashmap.set(sorted, []);
      hashmap.get(sorted).push(item);
    }
  }

  return hashmap.values();
}

console.log(groupAnagram1(input_value));

// 방법 2)

function groupAnagram2(strs) {
  hashmap = new Map();

  for (const item of strs) {
    const arr = Array(26).fill(0);

    for (const char of item) {
      // 아스키 코드 값을 이용
      arr[char.charCodeAt(0) - 'a'.charCodeAt(0)] += 1;
    }

    // ? 키가 배열일 경우 배열 요소가 같아도 주소가 다르기 때문에 동등 비교가 불가능
    const key = arr.toString();

    if (hashmap.has(key)) {
      hashmap.get(key).push(item);
    } else {
      hashmap.set(key, []);
      hashmap.get(key).push(item);
    }
  }

  return hashmap.values();
}

console.log(groupAnagram2(input_value));
